neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -0,0 +1,172 @@
+                           scope="query_proj_head{}".format(i)),
+                    linear(self.attention_keys, self._head_dim,
+                           scope="keys_proj_head{}".format(i)),
+                    linear(self.attention_values, self._head_dim,
+                           scope="values_proj_head{}".format(i)))
+                for i in range(self.n_heads)])
+
+            context = linear(tf.concat(head_contexts, -1), self._dimension,
+                             scope="output_proj")
+
+        next_contexts = loop_state.contexts.write(step, context)
+        next_head_weights = [loop_state.head_weights[i].write(step,
+                                                              head_weights[i])
+                             for i in range(self.n_heads)]
+
+        next_loop_state = MultiHeadLoopStateTA(
+            contexts=next_contexts,
+            head_weights=next_head_weights)
+
+        return context, next_loop_state
+
+    def attention_single_head(self, query: tf.Tensor,
+                              keys: tf.Tensor,
+                              values: tf.Tensor) -> Tuple[tf.Tensor,
+                                                          tf.Tensor]:
+        # shape: batch, time (similarities of attention keys in batch and time
+        # to the queries in the batch)
+        dot_product = tf.reduce_sum(
+            tf.expand_dims(query, 1) * keys, [-1])
+        energies = dot_product * self._scaling_factor
+
+        weights = tf.nn.softmax(energies)
+
+        if self.attention_mask is not None:
+            weights_all = weights * self.attention_mask
+            norm = tf.reduce_sum(weights_all, 1, keep_dims=True) + 1e-8
+            weights = weights_all / norm
+
+        # apply dropout to the weights (Attention Dropout)
+        weights = dropout(weights, self.dropout_keep_prob, self.train_mode)
+
+        # sum up along the time axis
+        context = tf.reduce_sum(
+            tf.expand_dims(weights, -1) * values, [1])
+
+        return context, weights
+
+    def initial_loop_state(self) -> MultiHeadLoopStateTA:
+        return MultiHeadLoopStateTA(
+            contexts=tf.TensorArray(
+                dtype=tf.float32, size=0, dynamic_size=True,
+                name="contexts"),
+            head_weights=[tf.TensorArray(
+                dtype=tf.float32, size=0, dynamic_size=True,
+                name="distributions_head{}".format(i), clear_after_read=False)
+                          for i in range(self.n_heads)])
+
+    def finalize_loop(self, key: str,
+                      last_loop_state: MultiHeadLoopStateTA) -> None:
+        for i in range(self.n_heads):
+            head_weights = last_loop_state.head_weights[i].stack()
+            self.histories["{}_head{}".format(key, i)] = head_weights
+
+    @property
+    def context_vector_size(self) -> int:
+        return self.attention_values.get_shape()[-1].value
+
+    def visualize_attention(self, key: str) -> None:
```

Comment:
Jak se to liší od base attentionu?

Reply:
je to celý o dimenzi větší. Za každou hlavu máš jednu historii. A to ještě přijde druhá dimenze za to, že v tom transformeru děláš attetnion pro několik queries současně.